File: apps/patter-voice-sdk/libraries/typescript/src/tools/mcp-client.ts
/**
 * Model Context Protocol (MCP) client integration for Patter.
 *
 * Lets users plug a Patter agent into MCP servers (Google Workspace,
 * PayPal, Postgres, GitHub, …) without writing a wrapper handler per
 * service:
 *
 *   ```ts
 *   phone.agent({
 *     mcpServers: [
 *       'https://mcp.googleworkspace.com/sse',
 *       { url: 'https://mcp.paypal.com/sse', headers: { Authorization: '...' } },
 *     ],
 *   });
 *   ```
 *
 * At call start, the SDK queries each server's ``tools/list``,
 * registers the discovered tools as Patter ``ToolDefinition``s with a
 * synthetic ``handler`` that dispatches to ``tools/call``, and merges
 * them into ``agent.tools`` before sending the function-tool list to
 * the underlying model (Realtime ``session.update`` or pipeline LLM).
 *
 * Lazy import: ``@modelcontextprotocol/sdk`` is an optional dependency.
 * Users who do not configure ``mcpServers`` never pay the install cost
 * and the SDK ships without a hard dependency.
 *
 * Limitations of the MVP (will iterate):
 *  - Per-call connection (handshake on every call: ~50-200 ms × N
 *    servers). Caching ``listTools`` results process-wide is a follow-up.
 *  - Streamable HTTP transport only — ``stdio`` and the legacy ``SSE``
 *    fallback are not exposed yet.
 *  - No tool-name conflict resolution: if an MCP tool collides with a
 *    user-defined ``agent.tools`` entry, MCP is rejected at startup.
 */

import type { ToolDefinition } from '../types';
import { getLogger } from '../logger';
import { validateWebhookUrl } from '../server';
import { VERSION } from '../version';

/** Public MCP server config. ``string`` is shorthand for ``{ url }``. */
export type MCPServerConfig =
  | string
  | {
      readonly url: string;
      /** Headers attached to every transport request — typically auth. */
      readonly headers?: Record<string, string>;
      /** Optional logical name for telemetry / log lines. */
      readonly name?: string;
    };

interface ResolvedConfig {
  readonly url: string;
  readonly headers: Record<string, string>;
  readonly name: string;
}

function resolveConfig(input: MCPServerConfig, index: number): ResolvedConfig {
  if (typeof input === 'string') {
    return { url: input, headers: {}, name: `mcp[${index}]` };
  }
  if (!input.url) {
    throw new Error(`mcpServers[${index}]: missing required 'url' field`);
  }
  return {
    url: input.url,
    headers: input.headers ?? {},
    name: input.name ?? `mcp[${index}]`,
  };
}

/** Loaded ``@modelcontextprotocol/sdk`` types — opaque to keep the
 *  optional dependency truly optional. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MCPClientLike = any;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MCPTransportLike = any;

interface ConnectedServer {
  readonly config: ResolvedConfig;
  readonly client: MCPClientLike;
  readonly transport: MCPTransportLike;
}

/**
 * Manages a set of MCP server connections for a single Patter call.
 * Lifecycle: ``connect()`` once, ``getDiscoveredTools()`` returns the
 * merged tool list with synthetic handlers, ``close()`` on call end.
 */
export class MCPManager {
  private readonly configs: ReadonlyArray<ResolvedConfig>;
  private connected: ConnectedServer[] = [];

  constructor(servers: ReadonlyArray<MCPServerConfig> | undefined) {
    this.configs = (servers ?? []).map((s, i) => resolveConfig(s, i));
  }

  get hasServers(): boolean {
    return this.configs.length > 0;
  }

  /** Connect to every configured server and discover their tools.
   *  Returns the discovered tools wrapped as Patter ``ToolDefinition``s. */
  async connect(): Promise<ToolDefinition[]> {
    if (this.configs.length === 0) return [];

    let mcpModule: { Client: new (info: { name: string; version: string }) => MCPClientLike };
    let transportModule: {
      StreamableHTTPClientTransport: new (
        url: URL,
        opts?: { requestInit?: { headers?: Record<string, string> } },
      ) => MCPTransportLike;
    };
    try {
      // Dynamic import keeps @modelcontextprotocol/sdk optional. Subpath
      // imports per the SDK's package.json exports map.
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      mcpModule = (await import('@modelcontextprotocol/sdk/client/index.js' as any)) as typeof mcpModule;
      transportModule = (await import(
        '@modelcontextprotocol/sdk/client/streamableHttp.js' as never
      )) as typeof transportModule;
    } catch (e) {
      throw new Error(
        'mcpServers configured but `@modelcontextprotocol/sdk` is not installed. ' +
          'Run `npm install @modelcontextprotocol/sdk` to enable MCP support. ' +
          `(import error: ${String(e)})`,
      );
    }

    const aggregatedTools: ToolDefinition[] = [];
    for (const cfg of this.configs) {
      try {
        validateWebhookUrl(cfg.url);
      } catch (e) {
        getLogger().error(`MCP server '${cfg.name}' (${cfg.url}) rejected by SSRF guard: ${String(e)}`);
        continue;
      }
      const transport = new transportModule.StreamableHTTPClientTransport(new URL(cfg.url), {
        requestInit: { headers: cfg.headers },
      });
      const client = new mcpModule.Client({ name: 'patter', version: VERSION });
      try {
        await client.connect(transport);
      } catch (e) {
        getLogger().error(`MCP server '${cfg.name}' (${cfg.url}) connect failed: ${String(e)}`);
        // Best effort cleanup of the half-connected transport.
        try { await transport.close?.(); } catch { /* ignore */ }
        continue;
      }
      this.connected.push({ config: cfg, client, transport });

      let listed: { tools?: Array<{ name: string; description?: string; inputSchema?: Record<string, unknown> }> };
      try {
        listed = await client.listTools();
      } catch (e) {
        getLogger().error(`MCP server '${cfg.name}' tools/list failed: ${String(e)}`);
        continue;
      }
      const tools = Array.isArray(listed?.tools) ? listed.tools : [];
      for (const t of tools) {
        if (!t?.name) continue;
        aggregatedTools.push({
          name: t.name,
          description: t.description ?? '',
          parameters: (t.inputSchema as Record<string, unknown>) ?? { type: 'object', properties: {} },
          handler: async (args: Record<string, unknown>): Promise<string> => {
            let callResult: { content?: Array<{ type: string; text?: string }>; isError?: boolean };
            try {
              callResult = await client.callTool({
                name: t.name,
                arguments: args,
              });
            } catch (err) {
              // Return the envelope instead of throwing: a thrown transport
              // error reached DefaultToolExecutor's retry loop, which would
              // re-fire a non-idempotent MCP tool (send_email…) up to 3
              // times on a transient error. Parity with the Python handler.
              return JSON.stringify({
                error: `MCP tool '${t.name}' failed: ${String(err).slice(0, 200)}`,
                fallback: true,
              });
            }
            const text = (callResult.content ?? [])
              .map((c) => (c.type === 'text' ? c.text ?? '' : JSON.stringify(c)))
              .join('\n');
            // MCP errors arrive as ``isError: true`` with a content
            // payload; surface them as the structured fallback shape so
            // the model can recover gracefully (parity with our local
            // tool error envelope).
            if (callResult.isError) {
              return JSON.stringify({ error: text || 'MCP tool error', fallback: true });
            }
            return text || '{}';
          },
        });
      }
      getLogger().info(`MCP server '${cfg.name}' registered ${tools.length} tool(s)`);
    }
    return aggregatedTools;
  }

  /** Validate no tool name collides between MCP-discovered and
   *  user-supplied tools. Throws on conflict so the user fixes it. */
  static assertNoConflicts(
    userTools: ReadonlyArray<ToolDefinition> | undefined,
    mcpTools: ReadonlyArray<ToolDefinition>,
  ): void {
    if (!userTools || userTools.length === 0 || mcpTools.length === 0) return;
    const userNames = new Set(userTools.map((t) => t.name));
    for (const mcp of mcpTools) {
      if (userNames.has(mcp.name)) {
        throw new Error(
          `MCP tool '${mcp.name}' collides with a user-supplied tool of the same name. ` +
            'Rename one of them or remove the duplicate from agent.tools.',
        );
      }
    }
  }

  /** Close every open MCP connection. Idempotent; logs but does not
   *  throw on individual failures (we don't want a flaky shutdown to
   *  derail the call-end teardown). */
  async close(): Promise<void> {
    const conns = this.connected;
    this.connected = [];
    for (const conn of conns) {
      try {
        await conn.client.close?.();
      } catch (e) {
        getLogger().debug(`MCP server '${conn.config.name}' close error (ignored): ${String(e)}`);
      }
    }
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/tools/mcp_client.py
"""Model Context Protocol (MCP) client integration for Patter (Python
parity with TS ``libraries/typescript/src/tools/mcp-client.ts``).

Lets users plug a Patter agent into MCP servers (Google Workspace,
PayPal, Postgres, GitHub, ...) without writing a wrapper handler per
service::

    phone.agent(
        ...,
        mcp_servers=[
            "https://mcp.googleworkspace.com/sse",
            {"url": "https://mcp.paypal.com/sse", "headers": {"Authorization": "..."}},
        ],
    )

At call start, the SDK queries each server's ``tools/list``, registers
the discovered tools with synthetic handlers that dispatch to
``tools/call``, and merges them into the agent's tool list before the
underlying model sees them.

Lazy import: ``mcp`` is an optional dependency declared in the
``[mcp]`` extra. Users who do not configure ``mcp_servers`` never pay
the install cost.

Limitations of the MVP (will iterate):

  * Per-call connection (handshake on every call). Caching the
    discovered tool list process-wide is a follow-up.
  * Streamable HTTP transport only — ``stdio`` and the legacy ``SSE``
    fallback are not exposed yet.
  * No tool-name conflict resolution: if an MCP tool collides with a
    user-supplied ``agent.tools`` entry, MCP is rejected at startup.
"""

from __future__ import annotations

import json
import logging
from typing import Any

logger = logging.getLogger("getpatter")

#: Public MCP server config. ``str`` is shorthand for ``{"url": <str>}``.
MCPServerConfig = str | dict


def _resolve_config(input_: MCPServerConfig, index: int) -> dict:
    if isinstance(input_, str):
        return {"url": input_, "headers": {}, "name": f"mcp[{index}]"}
    if not isinstance(input_, dict):
        raise TypeError(
            f"mcp_servers[{index}] must be a str URL or dict (got {type(input_).__name__})"
        )
    url = input_.get("url")
    if not url:
        raise ValueError(f"mcp_servers[{index}]: missing required 'url' field")
    return {
        "url": url,
        "headers": input_.get("headers") or {},
        "name": input_.get("name") or f"mcp[{index}]",
    }


class MCPManager:
    """Manages a set of MCP server connections for a single Patter call.

    Lifecycle: ``connect()`` once, returns the discovered tools as
    Patter tool dicts; ``close()`` on call end.
    """

    def __init__(self, servers: list[MCPServerConfig] | None) -> None:
        servers = servers or []
        self._configs = [_resolve_config(s, i) for i, s in enumerate(servers)]
        self._sessions: list[Any] = []
        self._exit_stacks: list[Any] = []

    @property
    def has_servers(self) -> bool:
        return len(self._configs) > 0

    async def connect(self) -> list[dict]:
        """Connect to every configured server and discover their tools.

        Returns a list of Patter-shape tool dicts with synthetic
        handlers wired to MCP ``tools/call``.
        """
        if not self._configs:
            return []

        try:
            from mcp import ClientSession  # type: ignore[import-not-found]
            from mcp.client.streamable_http import streamablehttp_client  # type: ignore[import-not-found]
        except ImportError as exc:
            raise RuntimeError(
                "mcp_servers configured but the `mcp` package is not "
                "installed. Run `pip install getpatter[mcp]` to enable "
                f"MCP support. (import error: {exc})"
            ) from exc

        from contextlib import AsyncExitStack

        from getpatter.tools.tool_executor import _validate_webhook_url

        aggregated: list[dict] = []
        for cfg in self._configs:
            # SSRF guard: refuse to connect to MCP servers on internal /
            # loopback / link-local / private targets (cloud metadata at
            # 169.254.169.254, localhost/127.0.0.0/8, 10/8, 172.16/12,
            # 192.168/16, ::1, …). Parity with the TS ``validateWebhookUrl``
            # check in mcp-client.ts. Best-effort (DNS rebinding can still
            # bypass) but MCP URLs are user-supplied config, not caller input.
            try:
                _validate_webhook_url(cfg["url"])
            except ValueError as exc:
                logger.error(
                    "MCP server '%s' (%s) rejected by SSRF guard: %s",
                    cfg["name"],
                    cfg["url"],
                    exc,
                )
                continue
            stack = AsyncExitStack()
            try:
                read_stream, write_stream, _ = await stack.enter_async_context(
                    streamablehttp_client(cfg["url"], headers=cfg["headers"])
                )
                session = await stack.enter_async_context(
                    ClientSession(read_stream, write_stream)
                )
                await session.initialize()
            except Exception as exc:  # noqa: BLE001
                logger.error(
                    "MCP server '%s' (%s) connect failed: %s",
                    cfg["name"],
                    cfg["url"],
                    exc,
                )
                await stack.aclose()
                continue
            self._sessions.append(session)
            self._exit_stacks.append(stack)

            try:
                listed = await session.list_tools()
            except Exception as exc:  # noqa: BLE001
                logger.error("MCP server '%s' tools/list failed: %s", cfg["name"], exc)
                continue
            tools = listed.tools if hasattr(listed, "tools") else []
            for t in tools:
                name = getattr(t, "name", None)
                if not name:
                    continue
                description = getattr(t, "description", "") or ""
                input_schema = getattr(t, "inputSchema", None) or {
                    "type": "object",
                    "properties": {},
                }
                aggregated.append(
                    _build_tool_entry(session, name, description, input_schema)
                )
            logger.info(
                "MCP server '%s' registered %d tool(s)", cfg["name"], len(tools)
            )
        return aggregated

    @staticmethod
    def assert_no_conflicts(
        user_tools: list[dict] | None, mcp_tools: list[dict]
    ) -> None:
        """Raise on tool-name collisions between user-supplied and
        MCP-discovered tools."""
        if not user_tools or not mcp_tools:
            return
        user_names = {t.get("name") for t in user_tools}
        for mcp_t in mcp_tools:
            if mcp_t.get("name") in user_names:
                raise ValueError(
                    f"MCP tool '{mcp_t['name']}' collides with a user-supplied "
                    "tool of the same name. Rename one of them or remove the "
                    "duplicate from agent.tools."
                )

    async def close(self) -> None:
        """Close every open MCP connection. Idempotent; logs but does
        not raise on individual failures."""
        stacks = self._exit_stacks
        self._exit_stacks = []
        self._sessions = []
        for stack in stacks:
            try:
                await stack.aclose()
            except Exception as exc:  # noqa: BLE001
                logger.debug("MCP close error (ignored): %s", exc)


def _build_tool_entry(
    session: Any, name: str, description: str, input_schema: dict
) -> dict:
    """Wrap an MCP tool descriptor into a Patter tool dict with a
    synthetic handler that dispatches to ``tools/call``."""

    async def _handler(args: dict, _ctx: dict) -> str:
        try:
            result = await session.call_tool(name, arguments=args)
        except Exception as exc:  # noqa: BLE001
            return json.dumps(
                {"error": f"MCP tool '{name}' error: {exc}", "fallback": True}
            )
        # MCP responses carry a ``content`` list of typed blocks.
        text_parts: list[str] = []
        content = getattr(result, "content", None) or []
        for block in content:
            block_type = getattr(block, "type", None)
            if block_type == "text":
                text_parts.append(getattr(block, "text", "") or "")
            else:
                # Non-text blocks are JSON-serialised verbatim so the
                # model still sees something useful.
                try:
                    text_parts.append(json.dumps(block.__dict__))
                except Exception:  # noqa: BLE001
                    text_parts.append(str(block))
        text = "\n".join(text_parts)
        if getattr(result, "isError", False):
            return json.dumps({"error": text or "MCP tool error", "fallback": True})
        return text or "{}"

    return {
        "name": name,
        "description": description,
        "parameters": input_schema,
        "handler": _handler,
    }
